support-figma/extended-layout-plugin/src/localization-module.ts

```diff
@@ -0,0 +1,274 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as PluginUtils from "./utils"
+
+const STRING_RES_PLUGIN_DATA_KEY = "vsw-string-res";
+
+interface StringResource {
+  name: string;
+  translatable: boolean;
+  // It can be a simple string or a string array.
+  text: string | string[];
+  textNodes: string[];
+}
+
+export async function generateLocalizationData(uploadedStrings: string[]) {
+  console.log("### Localization tool: generate string res name.");
+
+  await figma.loadAllPagesAsync();
+
+  let output = new Map<string, StringResource>();
+
+  // strings.xml files does not allow duplicates so no checks for duplicates here.
+  for (let uploadedString of uploadedStrings) {
+    let strRes = uploadedString as unknown as StringResource;
+    output.set(strRes.name, strRes);
+  }
+
+  for (let page of figma.root.children) {
+    for (let child of page.children) {
+      await localizeNodeAsync(child, output);
+    }
+  }
+
+  // Convert to an array of key-value pairs
+  const outputArray = Array.from(output);
+  figma.ui.postMessage({ msg: "localization-output", output: outputArray });
+}
+
+export async function updateStringResName(strRes: StringResource) {
+  console.log("### Localization tool: user renamed, update string res name.");
+  await figma.loadAllPagesAsync();
+
+  for (let nodeId of strRes.textNodes) {
+    const node = await figma.getNodeByIdAsync(nodeId);
+    if (node?.type == "TEXT") {
+      saveToPluginData(node, strRes.name);
+    } else {
+      console.error(
+        `### ${nodeId} is not a text node to update the string res name.`
+      );
+    }
+  }
+}
+
+export async function deleteStringResName(strRes: StringResource) {
+  console.log("### Localization tool: user removed, delete string res name.");
+  if (!strRes.name) {
+    console.error("### Expecting the resource name is set to empty");
+  }
+  strRes.name = "";
+  await updateStringResName(strRes);
+}
+
+async function localizeNodeAsync(
+  node: SceneNode,
+  output: Map<string, StringResource>
+) {
+  if (node.type == "TEXT") {
+    // Nodes with name starting with # is for local customization.
+    if (node.name.startsWith("#")) {
```

Comment:
Hmm we should have a discussion about this. There's no requirement or enforcement that customized text nodes in DesignCompose start with '#'. It's just a convention we've been using. @timothyfroehlich @iamralpht

Reply:
One possibility is to have these nodes show up in the plugin, but have the "translatable" column default to "false", and then have that field also be editable.

Reply:
I will leave it for v2.
I am actually thinking about uploading the app generated json file which contains those nodes that do client customization.